File: core/src/main/java/nl/sjtek/control/core/modules/Coffee.java
package nl.sjtek.control.core.modules;

import nl.sjtek.control.core.events.Bus;
import nl.sjtek.control.core.network.Arguments;
import nl.sjtek.control.data.ampq.events.LightEvent;
import nl.sjtek.control.data.responses.CoffeeResponse;
import nl.sjtek.control.data.responses.Response;

import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/**
 * Created by wouter on 14-11-16.
 */
public class Coffee extends BaseModule {

    private static final long HEAT_DELAY = 600000;
    private final ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1);
    private long lastTriggered;

    public Coffee(String key) {
        super(key);
    }

    public synchronized void start(Arguments arguments) {
        if (!isHeated()) {
            new EnableThread().start();
        }
    }

    @Override
    public Response getResponse() {
        return new CoffeeResponse(lastTriggered, isHeated());
    }

    private boolean isHeated() {
        return (System.currentTimeMillis() - lastTriggered) < HEAT_DELAY;
    }

    @Override
    public String getSummaryText() {
        return "The coffee is " + (isHeated() ? "heated." : "not heated.");
    }

    @Override
    public boolean isEnabled(String user) {
        return false;
    }

    private class EnableThread extends Thread {
        @Override
        public void run() {
            super.run();
            Bus.post(new LightEvent(10, false));
            try {
                Thread.sleep(3000);
            } catch (InterruptedException ignored) {
            }
            Bus.post(new LightEvent(10, true));
            lastTriggered = System.currentTimeMillis();
            dataChanged(true);
            executor.schedule(new Runnable() {
                @Override
                public void run() {
                    dataChanged(true);
                }
            }, HEAT_DELAY + 100, TimeUnit.MILLISECONDS);
        }
    }
}

File: core/src/main/kotlin/nl/sjtek/control/core/modules/Coffee.kt
package nl.sjtek.control.core.modules

import net.engio.mbassy.listener.Handler
import nl.sjtek.control.core.events.Bus
import nl.sjtek.control.core.events.CoffeeEvent
import nl.sjtek.control.core.events.SwitchEvent
import nl.sjtek.control.core.get
import nl.sjtek.control.data.response.Coffee
import nl.sjtek.control.data.response.Response
import spark.Spark.path
import java.util.*
import kotlin.concurrent.schedule

class Coffee(key: String) : Module(key) {
    private val TIMEOUT: Long = 600000
    private val enabled: Boolean
        get() = System.currentTimeMillis() - lastTimeEnabled < TIMEOUT
    private var lastTimeEnabled: Long = 0
    override val response: Response
        get() = Coffee(key, enabled, lastTimeEnabled)

    init {
        Bus.subscribe(this)
    }

    override fun initSpark() {
        path("/coffee") {
            get("/enable", this::enable)
        }
    }

    @Handler
    fun onCoffeeEvent(event: CoffeeEvent) = enable()

    private fun enable(req: spark.Request? = null, res: spark.Response? = null) {
        if (!enabled) {
            lastTimeEnabled = System.currentTimeMillis()
            Bus.post(SwitchEvent(10, false))
            Timer().schedule(2000) {
                Bus.post(SwitchEvent(10, true))
            }
        }
    }
}
